src/main.rs

```diff
@@ -242,5 +279,93 @@ fn run() -> FeludaResult<()> {
     }
 
     log(LogLevel::Info, "Feluda completed successfully");
+
+    Ok(())
+}
+
+fn validate_ssh_key(key_path: &Path) -> Result<(), git2::Error> {
+    if !key_path.exists() {
+        log(LogLevel::Error, &format!("SSH key file not found: {}", key_path.display()));
+        return Err(git2::Error::from_str("SSH key file not found"));
+    }
+    if key_path.extension().map(|ext| ext == "pub").unwrap_or(false) {
+        log(LogLevel::Error, &format!("Invalid SSH key: {} is a public key (.pub)", key_path.display()));
+        return Err(git2::Error::from_str("Public key provided instead of private key"));
+    }
     Ok(())
 }
+
+fn clone_repository(repo_url: &str, dest_path: &Path, token: Option<&str>, ssh_key: Option<&str>, ssh_passphrase: Option<&str>) -> FeludaResult<()> {
```

Comment:
Please also write tests for these functions, cause they are new function additions.